multi/grlib-multi/spacewire.c

```diff
@@ -0,0 +1,828 @@
+
+	TRACE("nPackets: %d", nPackets);
+
+	/* Setup descriptors */
+	size_t firstDesc = dev->lastTxDesc;
+	const size_t lastDesc = (dev->lastTxDesc + nPackets) % SPW_TX_DESC_CNT;
+	bool wrapped = (lastDesc <= firstDesc);
+
+	for (size_t cnt = 0; cnt < nPackets; cnt++) {
+		(void)mutexLock(dev->txIrqLock);
+		while ((dev->txDescFree == 0) || dev->txWaited[dev->lastTxDesc]) {
+			/* Wait for free descriptor or for packet to be acknowledged */
+			(void)condWait(dev->cond, dev->txIrqLock, 0);
+		}
+		dev->txDescFree--;
+		(void)mutexUnlock(dev->txIrqLock);
+
+		volatile spw_txDesc_t *desc = &dev->txDesc[dev->lastTxDesc];
+
+		spw_txPacket_t packet;
+		buf += spw_txMsgToPacket(buf, &packet);
+		/* Interrupt after each packet transmitted */
+		desc->ctrl = (packet.flags & TX_DESC_USR_MSK) | TX_DESC_IE;
+		if (dev->lastTxDesc == SPW_TX_DESC_CNT - 1) {
+			/* Wrap around */
+			desc->ctrl |= TX_DESC_WR;
+		}
+		desc->hdrAddr = (void *)va2pa((void *)packet.hdr);
+		desc->packetLen = packet.dataLen;
+		desc->dataAddr = (void *)va2pa((void *)packet.data);
+
+		/* Everything is set up, enable descriptor */
+		desc->ctrl |= TX_DESC_EN;
+
+		dev->txWaited[dev->lastTxDesc] = true;
+
+		/* Start transmission */
+		dev->vbase[DMA_CTRL] |= DMA_CTRL_TE;
+
+		dev->lastTxDesc = (dev->lastTxDesc + 1) % SPW_TX_DESC_CNT;
+	}
+
+	TRACE("Packets set up");
+
+	/* Wait for transmission to finish */
+	while ((firstDesc <= lastDesc) || (wrapped && (firstDesc > lastDesc))) {
+		if ((dev->txDesc[firstDesc].ctrl & TX_DESC_EN) == 0) {
+			dev->txWaited[firstDesc] = false;
+			size_t next = (firstDesc + 1) % SPW_TX_DESC_CNT;
+			if ((next == 0) && wrapped) {
+				wrapped = false;
+			}
+			firstDesc = next;
+		}
+		else {
+			(void)condWait(dev->cond, dev->txLock, 0);
+		}
+	}
+
+	TRACE("Packets sent");
+
+	(void)mutexUnlock(dev->txLock);
+
+	return nPackets;
+}
+
+
+/* Starts transmission and returns */
+static int spw_transmitAsync(spw_dev_t *dev, const uint8_t *buf, const size_t nPackets)
+{
+	(void)mutexLock(dev->txLock);
+
+	TRACE("nPackets: %d", nPackets);
+
+	/* Setup descriptors */
+	for (size_t cnt = 0; cnt < nPackets; cnt++) {
+		(void)mutexLock(dev->txIrqLock);
+		while ((dev->txDescFree == 0) || dev->txWaited[dev->lastTxDesc]) {
+			/* Wait for free descriptor or for packet to be acknowledged */
+			(void)condWait(dev->cond, dev->txIrqLock, 0);
+		}
+		dev->txDescFree--;
+		(void)mutexUnlock(dev->txIrqLock);
+
+		volatile spw_txDesc_t *desc = &dev->txDesc[dev->lastTxDesc];
+
+		spw_txPacket_t packet;
+		buf += spw_txMsgToPacket(buf, &packet);
+		/* Interrupt after each packet transmitted */
+		desc->ctrl = (packet.flags & TX_DESC_USR_MSK) | TX_DESC_IE;
+		if (dev->lastTxDesc == SPW_TX_DESC_CNT - 1) {
+			/* Wrap around */
+			desc->ctrl |= TX_DESC_WR;
+		}
+		desc->packetLen = packet.dataLen;
+
+		uint8_t hdrLen = packet.flags & TX_DESC_HDR_LEN;
+		volatile void *txBuff = dev->txBuff[dev->lastTxDesc];
+
+		memcpy((void *)txBuff, packet.hdr, hdrLen);
```

Comment:
Not the best practice to memcpy to volatile buff, doesn't matter in this case I guess

Reply:
I agree, here volatile doesn't apply just yet